### PR TITLE
Apply clipping and translation transforms correctly

### LINKED ISSUE #3975: Cannot change the BackColor of ToolStrip control
* .NET Core Version: 
.NET 5.0.100-rc.2.20471.13

* Have you experienced this same bug with .NET Framework?: 
No

**More Info:**

1. This is a regression issue, it does not occur in .NET SDK 5.0.100-Preview.8.20417.9, it start from .NET 5.0 RC1.
2. This issue also can reproduce in the Properties window(design mode). 

**Problem description:**

Cannot change the BackColor of ToolStrip control
core

**Expected behavior:**

The BackColor of ToolStrip control can be changed correctly. 
framework

**Minimal repro:**
[Testapp.zip](https://github.com/dotnet/winforms/files/5260320/Testapp.zip)
1. Launch the attached project in VS and run it.
2. Try to change the BackColor property in PropertyGrid and check the result.

### PR DESCRIPTION
**Pull request by JeremyKuhne**
Logic was off for some of the cases where we set up an `HDC` for drawing and we weren't applying clipping or translation transforms when we needed to. For the `ToolStrip` this causes the background of the parent to fill over the `ToolStrip` background.

This has implications for most rendering where a clip or transform is applied.

Add tests for the `HDC` handling struct to validate application of `Graphics` properties.

Fixes #3975

## Proposed changes

- Fix logic in `DeviceContextHDCScope`
- Add tests for property application

## Customer Impact

- Background color isn't visible in Toolstrip (possibly other internal controls, testing hasn't surfaced any others yet)
- Customer or control vendor's controls that utilize clipping or origin transforms may not render correctly if they utilize ControlPaint or base class rendering

## Regression? 

- Yes

## Risk

- Low

### Before

Visual styles enabled and disabled; Professional and ManagerRenderMode RenderMode
image

### After

Visual styles enabled and disabled; Professional and ManagerRenderMode RenderMode
image

`RenderMode.System` (not the default) was validated to be unimpacted by this change.

## Test methodology 

- Manually validated screen shots for various permutations
- Wrote regression tests for HDC/Graphics property application

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/3990)

### CODE REVIEW

File: src/System.Windows.Forms.Primitives/src/System/Windows/Forms/DeviceContextHdcScope.cs [hunk 1]
```diff
@@ -83,41 +84,65 @@ internal class DeviceContextHdcScope : DisposalTracking.Tracker, IDisposable
...
-                }
-                else
-                {
-                    // We have a provider
-                    HDC = provider.GetHDC();
+                // We have an IDeviceContext (case 3 above), we can't apply properties because there is no
+                // Graphics object available.
+                needToApplyProperties = false;
+            }
+            else if (provider != null && provider.IsGraphicsStateClean)
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms.Primitives/src/System/Windows/Forms/DeviceContextHdcScope.cs:108]: provider?.IsGraphicsStateClean

File: src/System.Windows.Forms.Primitives/src/System/Windows/Forms/DeviceContextHdcScope.cs [hunk 2]
```diff
@@ -83,41 +84,65 @@ internal class DeviceContextHdcScope : DisposalTracking.Tracker, IDisposable
...
-                    if (HDC.IsNull)
+            if (provider != null)
+            {
+                // We have a provider, grab the underlying HDC if possible unless we know we've created and
+                // modified a Graphics object around it.
+                HDC = needToApplyProperties ? default : provider.GetHDC();
+
+                if (HDC.IsNull)
+                {
+                    graphics = provider.GetGraphics(createIfNeeded: true);
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms.Primitives/src/System/Windows/Forms/DeviceContextHdcScope.cs:122]: I see that `createIfNeeded` was true before your change, but I don't understand the scenario, we know that properties were updated on the graphics object, but the object does not exist?

File: src/System.Windows.Forms.Primitives/src/System/Windows/Forms/DeviceContextHdcScope.cs [hunk 3]
```diff
@@ -83,41 +84,65 @@ internal class DeviceContextHdcScope : DisposalTracking.Tracker, IDisposable
...
-                        DeviceContext = graphics;
+                        throw new InvalidOperationException();
                     }
+                    DeviceContext = graphics;
                 }
+            }
 
+            if (!needToApplyProperties || graphics is null)
+            {
+                HDC = !HDC.IsNull ? HDC : (Gdi32.HDC)DeviceContext.GetHdc();
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms.Primitives/src/System/Windows/Forms/DeviceContextHdcScope.cs:133]: I think double negation is hard to read.
                HDC = HDC.IsNull ? (Gdi32.HDC)DeviceContext.GetHdc() : HDC;

File: src/System.Windows.Forms.Primitives/src/System/Windows/Forms/DeviceContextHdcScope.cs [hunk 4]
```diff
@@ -83,41 +84,65 @@ internal class DeviceContextHdcScope : DisposalTracking.Tracker, IDisposable
...
+                Debug.Assert(!HDC.IsNull);
                 _savedHdcState = saveHdcState ? Gdi32.SaveDC(HDC) : 0;
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms.Primitives/src/System/Windows/Forms/DeviceContextHdcScope.cs:135]: Should we throw if HDC is null? Maybe in 6.0 and not in 5.0.

**Review by Tanya-Solyanik — APPROVED** [review]
Nice tests!

File: src/System.Windows.Forms.Primitives/src/System/Windows/Forms/DeviceContextHdcScope.cs (hunk 1 above)
Comment (JeremyKuhne) [reply src/System.Windows.Forms.Primitives/src/System/Windows/Forms/DeviceContextHdcScope.cs:108]: You would have to do `provider?.IsGraphicsStateClean ?? false`, which seemed a little less optimal than what I went for.

File: src/System.Windows.Forms.Primitives/src/System/Windows/Forms/DeviceContextHdcScope.cs (hunk 2 above)
Comment (JeremyKuhne) [reply src/System.Windows.Forms.Primitives/src/System/Windows/Forms/DeviceContextHdcScope.cs:122]: We _should_ be able to say false in this case, but it is a little less resilient to other mistakes and awkward to think through either way so I went with true here.

File: src/System.Windows.Forms.Primitives/src/System/Windows/Forms/DeviceContextHdcScope.cs (hunk 3 above)
Comment (JeremyKuhne) [reply src/System.Windows.Forms.Primitives/src/System/Windows/Forms/DeviceContextHdcScope.cs:133]: Fair enough.

File: src/System.Windows.Forms.Primitives/src/System/Windows/Forms/DeviceContextHdcScope.cs (hunk 4 above)
Comment (JeremyKuhne) [reply src/System.Windows.Forms.Primitives/src/System/Windows/Forms/DeviceContextHdcScope.cs:135]: I actually need to remove the assert as some other tests push in with this as null.

File: src/System.Windows.Forms.Primitives/src/System/Windows/Forms/DeviceContextHdcScope.cs (hunk 1 above)
Comment (RussKie) [reply src/System.Windows.Forms.Primitives/src/System/Windows/Forms/DeviceContextHdcScope.cs:108]: You could also say: `provider?.IsGraphicsStateClean == true`

**Comment by JeremyKuhne** [comment]
Release failing for #3992 and #3856

**Review by Tanya-Solyanik — APPROVED** [review]

**Pull request merged**